### PR TITLE
travis: enable fuzzer regression tests under sanitizers

### PR DESCRIPTION
**Pull request by mrc0mmand**
Followup to #15877.

### CODE REVIEW

**Comment by evverx** [comment]
Looks like the gcc-asan-ubsan job was bordering on timing out :-) I'll restart it to see where it gets us.

**Comment by mrc0mmand** [comment]
> Looks like the gcc-asan-ubsan job was bordering on timing out :-) I'll restart it to see where it gets us.

I was just about to force push this PR, as in CentOS CI `--timeout-multiplier=4` is not enough for `fuzz-unit-file`, yet here it passes with `--timeout-multiplier=3` :-)

**Comment by mrc0mmand** [comment]
Either the clang run is suspiciously fast or the gcc one is suspiciously slow, becase:

clang
```
726/758 fuzz-unit-file:oss-fuzz-10007:address,undefined                                             OK             2.58s
727/758 fuzz-unit-file:oss-fuzz-11569:address,undefined                                             OK             3.43s
728/758 fuzz-unit-file:oss-fuzz-13125:address,undefined                                             OK             2.73s
```

gcc
```
728/760 fuzz-unit-file:oss-fuzz-10007:address,undefined                                             OK             112.69s
729/760 fuzz-unit-file:oss-fuzz-11569:address,undefined                                             OK             115.17s
730/760 fuzz-unit-file:oss-fuzz-13125:address,undefined                                             OK             88.85s
```

and for complete comparison current CentOS CI (gcc):
```
728/760 fuzz-unit-file:oss-fuzz-10007:address,undefined                                             OK             183.79s
729/760 fuzz-unit-file:oss-fuzz-11569:address,undefined                                             TIMEOUT        300.04s
730/760 fuzz-unit-file:oss-fuzz-13125:address,undefined                                             OK             143.22s
```

**Comment by mrc0mmand** [comment]
I suspect this might be caused by meson adding `-O0` to gcc runs implicitly, but in the clang run there is no optimization level on the cc command line. Either way, I think the sanitizer builds should be compiled with `--optimization=1` at least.

**Comment by mrc0mmand** [comment]
Hm, that wasn't it, even with -O1 the gcc run is significantly slower and the same situation happens in CentOS CI as well - the clang run is blazing fast whereas the gcc one timeouts.

**Comment by evverx** [comment]
Looking at https://github.com/systemd/systemd/blob/master/test/fuzz/meson.build (where `meson_build_sh` is run to build the fuzz targets) I'm starting to think that --optiminiztion=1 wasn't passed to `meson`. Were the fuzz targets really built with -O1?

**Comment by mrc0mmand** [comment]
> Looking at https://github.com/systemd/systemd/blob/master/test/fuzz/meson.build (where `meson_build_sh` is run to build the fuzz targets) I'm starting to think that --optiminiztion=1 wasn't passed to `meson`. Were the fuzz targets really built with -O1?

Ah, I see, it doesn't use the verbose flag, so the compilation flags are hidden, thus non-greppable. So that might be the culprit :-)

**Comment by mrc0mmand** [comment]
Definitely something's fishy going on. I tried both gcc & clang with the last commit locally, with following results:

gcc:
```
$ time build-gcc/fuzz-unit-file\:address\,undefined test/fuzz/fuzz-unit-file/oss-fuzz-11569 
test/fuzz/fuzz-unit-file/oss-fuzz-11569... ok

real    3m22.804s
user    3m18.725s
sys     0m0.245s
```

clang:
```
$ time build-clang/fuzz-unit-file\:address\,undefined test/fuzz/fuzz-unit-file/oss-fuzz-11569 
test/fuzz/fuzz-unit-file/oss-fuzz-11569... ok

real    0m3.222s
user    0m3.104s
sys     0m0.089s
```

This reminded me of https://github.com/systemd/systemd/issues/12997 and indeed, `detect_stack_use_after_return=0` cuts the time down to ~3 seconds on gcc. I'll need to look into this further...

**Comment by mrc0mmand** [comment]
This looks *suspiciously* similar to the #12997 issue, at least according to some perf-ing:

gcc:
```
# Samples: 862K of event 'cycles:u'
# Event count (approx.): 873078595486
#
# Overhead  Command          Shared Object             Symbol                                   
# ........  ...............  ........................  .........................................
#
    95.14%  fuzz-unit-file:  libasan.so.6.0.0          [.] __asan_stack_malloc_0
     3.13%  fuzz-unit-file:  libasan.so.6.0.0          [.] __asan_stack_malloc_1
     0.18%  fuzz-unit-file:  libc-2.31.so              [.] __strlen_avx2
     0.17%  fuzz-unit-file:  libasan.so.6.0.0          [.] __asan_stack_malloc_2
     0.15%  fuzz-unit-file:  libasan.so.6.0.0          [.] __asan_region_is_poisoned
     0.07%  fuzz-unit-file:  libc-2.31.so              [.] __strchr_avx2
     0.06%  fuzz-unit-file:  libsystemd-shared-245.so  [.] utf8_encoded_valid_unichar
...
```

clang:
```
# Samples: 9K of event 'cycles:u'
# Event count (approx.): 9432593211
#
# Overhead  Command          Shared Object             Symbol                                                                                                                                                                            
# ........  ...............  ........................  ..................................................................................................................................................................................
#
    14.31%  fuzz-unit-file:  libc-2.31.so              [.] __strlen_avx2
    13.24%  fuzz-unit-file:  fuzz-unit-file            [.] __sanitizer::mem_is_zero
     6.10%  fuzz-unit-file:  libsystemd-shared-245.so  [.] sipround
     5.21%  fuzz-unit-file:  libsystemd-shared-245.so  [.] siphash24_compress
     5.20%  fuzz-unit-file:  fuzz-unit-file            [.] __asan_region_is_poisoned
     3.91%  fuzz-unit-file:  fuzz-unit-file            [.] __asan_stack_malloc_0
     3.77%  fuzz-unit-file:  fuzz-unit-file            [.] __interceptor_memchr.part.0
```

**Comment by mrc0mmand** [comment]
perf with call graph:

image

I also tried to compile everything with -O2 (and gcc) to no avail.

**Comment by mrc0mmand** [comment]
Reported to gcc upstream as https://gcc.gnu.org/bugzilla/show_bug.cgi?id=95275.

File: test/fuzz/meson.build [hunk 1]
```diff
@@ -7,7 +7,7 @@ sanitize_address_undefined = custom_target(
                    project_source_root,
                    '@OUTPUT@',
                    'fuzzers',
-                   '-Db_lundef=false -Db_sanitize=address,undefined',
+                   '-Db_lundef=false -Db_sanitize=address,undefined --optimization=1',
```
Comment (evverx) [inline test/fuzz/meson.build]: I wonder if it would be better to replace it with `'... --optimization=@0@'.format(get_option('optimization'))`.

Comment (mrc0mmand) [reply test/fuzz/meson.build]: That's unquestionably better, thanks. I really need to get into meson properly anytime soon :-)

**Comment by anitazha** [comment]
Hey @mrc0mmand, I'm getting `test/fuzz/meson.build:3:0: ERROR:  Tried to access unknown option "optimization".` when building with meson 0.47.2. Seems like optimization was added in meson 0.48.0. Not sure if we want to bump the systemd requirement from meson >= 0.46 or do something else?

**Comment by mrc0mmand** [comment]
@anitazha hey! Sorry about that, unfortunately all CI's use a newer version, so it wasn't caught there. #15937 should address this.

**Comment by mrc0mmand** [comment]
@evverx just a heads up that fix for https://gcc.gnu.org/bugzilla/show_bug.cgi?id=94910 got into gcc-10 and gcc-9 branches, so we should be able to re-enable regression fuzzers under gcc soon™.

**Pull request merged**